driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -492,23 +492,51 @@ void setKeyspace(String keyspace) throws ConnectionException {
             return thisFuture;
         }
 
-        logger.trace("{} Setting keyspace {}", this, keyspace);
-        // Note: we quote the keyspace below, because the name is the one coming from Cassandra, so it's in the right case already
-        Future future = write(new Requests.Query("USE \"" + keyspace + '"'));
-        return Futures.transform(future, new Function<Message.Response, Connection>() {
-            @Override
-            public Connection apply(Message.Response response) {
-                if (response instanceof SetKeyspace) {
-                    Connection.this.keyspace = ((SetKeyspace) response).keyspace;
-                    return Connection.this;
-                } else if (response.type == ERROR) {
-                    Responses.Error error = (Responses.Error) response;
-                    throw defunct(error.asException(address));
-                } else {
-                    throw defunct(new DriverInternalError("Unexpected response while setting keyspace: " + response));
+        SetKeyspaceAttempt attempt = new SetKeyspaceAttempt(keyspace);
+
+        // Check for an existing keyspace attempt.
+        while (true) {
+            SetKeyspaceAttempt existingAttempt = targetKeyspace.get();
+            // if existing attempts' keyspace matches what we are trying to set, use it.
+            if (attempt.equals(existingAttempt)) {
+                attempt = existingAttempt;
+                break;
+            } else if (targetKeyspace.compareAndSet(existingAttempt, attempt)) {
```

Comment:
I agree, if the CAS fails, the new attempt is likely for the same keyspace, in that case the next iteration of the while loop should see the new `existingAttempt` has the same keyspace as what we are trying to set (line 501-502).   It isn't clear but the equals() on `SetKeyspaceAttempt` is on the keyspace (maybe I should just compare the keyspaces of the attempts so it is more clear?)

If the CAS succeeds and the existingAttempt is for a different keyspace, I think that's ok.  I added a debug statement (at 508) if the existingAttempt is still in flight, as this could cause some requests to fail because the keyspace is changing, as you said there is not much we can do about that.  Although maybe we should change that logging statement to warning so a user knows that this is happening, wdyt?